### PR TITLE
Gcc 8 fixes

### LINKED ISSUE #12984: GCC 8
## Rationale

MOOSE should work with GCC 8.  Currently has a few compile warnings:  #12807

## Description

See above

## Impact

None - just fixing compiler warnings

### PR DESCRIPTION
**Pull request by andrsd**
GCC 8.2.x barfs some warnings, but when we treat warnings as error the build fails.

closes #12984

### CODE REVIEW

**Comment by andrsd** [comment]
Tagging @milljm

**Comment by andrsd** [comment]
There is still one warning to fix: `NonlinearSystem.C:343`, it is this line:
```
MatFDColoringSetFunction(_fdcoloring,
                           (PetscErrorCode(*)(void)) & libMesh::libmesh_petsc_snes_fd_residual,
                           &petsc_nonlinear_solver);
```
I currently commented it out. The compiler complains about incompatibility between the PETSc prototype and libmesh prototype.
```
error: cast between incompatible function types from ‘PetscErrorCode (*)(SNES, Vec, Vec, void*)’ {aka ‘int (*)(_p_SNES*, _p_Vec*, _p_Vec*, void*)’} to ‘PetscErrorCode (*)()’ {aka ‘int (*)()’} [-Werror=cast-function-type]
                            (PetscErrorCode(*)(void)) & libMesh::libmesh_petsc_snes_fd_residual,
                                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Maybe @fdkong have a quick solution for this?

File: framework/src/postprocessors/FindValueOnLine.C [hunk 1]
```diff
@@ -90,7 +90,7 @@ FindValueOnLine::execute()
 
   bool found_it = false;
   Real value = 0;
-  for (auto i = decltype(_depth)(0); i < _depth; ++i)
+  for (unsigned int i = 0; i < _depth; ++i)
```
Comment (permcody) [inline framework/src/postprocessors/FindValueOnLine.C]: Interesting, so we used that syntax precisely to get around the `const` problem (as opposed to `for (decltype(_depth) i = 0 ...`. So I guess C++ really can't give us a good way to get the right type while removing const?

Comment (andrsd) [reply framework/src/postprocessors/FindValueOnLine.C]: yeah. I think the const on the member variable has value, but then removing the const after `decltype` is IMO much worse then just doing plain old `unsigned int i = ...`.

**Comment by fdkong** [comment]
> There is still one warning to fix: `NonlinearSystem.C:343`, it is this line:
> 
> ```
> MatFDColoringSetFunction(_fdcoloring,
>                            (PetscErrorCode(*)(void)) & libMesh::libmesh_petsc_snes_fd_residual,
>                            &petsc_nonlinear_solver);
> ```
> I currently commented it out. The compiler complains about incompatibility between the PETSc prototype and libmesh prototype.
> 
> ```
> error: cast between incompatible function types from ‘PetscErrorCode (*)(SNES, Vec, Vec, void*)’ {aka ‘int (*)(_p_SNES*, _p_Vec*, _p_Vec*, void*)’} to ‘PetscErrorCode (*)()’ {aka ‘int (*)()’} [-Werror=cast-function-type]
>                             (PetscErrorCode(*)(void)) & libMesh::libmesh_petsc_snes_fd_residual,
>                                                                  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ```
> Maybe @fdkong have a quick solution for this?

Calling @jedbrown

**Comment by andrsd** [comment]
Just for the record, this is gcc-8.2.0 and PETSc 3.9.4

File: framework/src/postprocessors/FindValueOnLine.C (hunk 1 above)
Comment (permcody) [reply framework/src/postprocessors/FindValueOnLine.C]: Agreed, but we've seen several places where `unsigned int` bites us as we grow beyond the 32-bit sizes. I wish there was a clean syntax that let us loop over the current size type.

I'm fine with this for now until we figure out something better.

**Comment by andrsd** [comment]
My temporary workaround is to add `-Wno-cast-function-type` to CPPFLAGS until we find a better solution.

**Comment by jedbrown** [comment]
Why this cast in the first place? The error says what type is needed, and hopefully that's the type of the actual function you're passing.

**Comment by andrsd** [comment]
Not that I am author of this code, but I guess it can stem from the following. I PETSs 3.9.2, I see:
```
dmdasnes.c:58
static PetscErrorCode SNESComputeFunction_DMDA(SNES snes,Vec X,Vec F,void *ctx) {...}
```

**Comment by jedbrown** [comment]
Oh, this is a funny function that can be called without a SNES. You can avoid that error with the ugly double cast

```
MatFDColoringSetFunction(matfdcoloring,(PetscErrorCode (*)(void))(void(*)(void))FormFunctionLocal,&user);
```

From the gcc man page:
> The function type `void (*) (void)` is special and matches everything, which can be used to suppress this warning.

File: test/Makefile [hunk 2]
```diff
@@ -9,7 +9,7 @@
 ###############################################################################
 MOOSE_DIR          ?= $(shell dirname `pwd`)
 FRAMEWORK_DIR      ?= $(MOOSE_DIR)/framework
-ADDITIONAL_CPPFLAGS += -Wall -Wextra
+ADDITIONAL_CPPFLAGS += -Wall -Wextra -Werror -Wno-cast-function-type
```
Comment (friedmud) [inline test/Makefile]: This doesn't work with older compilers (and probably isn't the right way to fix this anyway).

Comment (andrsd) [reply test/Makefile]: That's correct. It was not meant as a fix, but to get further in build to see what other issues we bump into. So I put it in a separate patch so we could drop it from the branch easily when we have a proper fix.

The next problem I had IIRC was some linking fail in exodiff. Could have been just my machine - I did not get to investigate it further. Just for the record if somebody is going to pick up this PR and continue the work.

**Review by permcody — APPROVED** [review]

**Pull request merged**